File: 大数据隐私保护/dsjclass02/main.go
package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"github.com/fernet/fernet-go"
	"io/ioutil"
	"log"
	"os"
	"strings"
)

type SSE struct {
	K1        []byte            // 密钥 K1
	K2        []byte            // 密钥 K2
	K3        []byte            // 密钥 K3
	ArrayA    map[int][]byte    // 存储加密节点的数组
	TableT    map[int][3]interface{} // 查找表
}

func NewSSE() *SSE {
	return &SSE{
		ArrayA: make(map[int][]byte),
		TableT: make(map[int][3]interface{}),
	}
}

// Setup算法：生成密钥
func (sse *SSE) Setup() {
	sse.K1 = generateKey()
	sse.K2 = generateKey()
	sse.K3 = generateKey()
}

// 生成一个新的密钥
func generateKey() []byte {
	key := make([]byte, 32)
	_, err := os.ReadFile("/dev/urandom", key)
	if err != nil {
		log.Fatal(err)
	}
	return key
}

// 伪随机置换函数
func (sse *SSE) PseudoRandomPerm(key []byte, value string) int {
	combined := append(key, []byte(value)...)
	hash := sha256.Sum256(combined)
	return int(binary.BigEndian.Uint32(hash[:4])) % 10000
}

// 伪随机函数
func (sse *SSE) PseudoRandomFunc(key []byte, value string) []byte {
	combined := append(key, []byte(value)...)
	hash := sha256.Sum256(combined)
	return hash[:]
}

// BuildIndex算法：构建索引
func (sse *SSE) BuildIndex(documents map[string]map[string]bool) {
	ctr := 1
	for keyword, docIDs := range documents {
		docIDList := make([]string, 0, len(docIDs))
		for docID := range docIDs {
			docIDList = append(docIDList, docID)
		}
		KiPrev := generateKey()

		// 随机生成 gamma 和 eta
		gamma := sse.PseudoRandomPerm(sse.K3, keyword)
		eta := sse.PseudoRandomFunc(sse.K2, keyword)

		firstAddr := sse.PseudoRandomPerm(sse.K1, fmt.Sprintf("%d", ctr))
		sse.TableT[gamma] = [3]interface{}{firstAddr, KiPrev, eta}

		for j, docID := range docIDList {
			KiCurr := generateKey()
			var nextAddr int
			if j < len(docIDList)-1 {
				nextAddr = sse.PseudoRandomPerm(sse.K1, fmt.Sprintf("%d", ctr+1))
			}

			node := map[string]interface{}{
				"doc_id":   docID,
				"next_key": KiCurr,
				"next_addr": nextAddr,
			}

			// 加密节点
			fernetKey := fernet.Key(KiPrev)
			encryptedNode, err := fernetKey.Encrypt([]byte(fmt.Sprintf("%v", node)))
			if err != nil {
				log.Fatal(err)
			}

			sse.ArrayA[sse.PseudoRandomPerm(sse.K1, fmt.Sprintf("%d", ctr))] = encryptedNode

			KiPrev = KiCurr
			ctr++
		}
	}
}

// GenTrapdoor算法：生成陷门
func (sse *SSE) GenTrapdoor(keyword string) (int, []byte) {
	gamma := sse.PseudoRandomPerm(sse.K3, keyword)
	eta := sse.PseudoRandomFunc(sse.K2, keyword)
	return gamma, eta
}

// Search算法：搜索文档
func (sse *SSE) Search(trapdoor [2]interface{}) []string {
	gamma := trapdoor[0].(int)
	eta := trapdoor[1].([]byte)

	var result []string
	if val, exists := sse.TableT[gamma]; exists {
		addr, key, storedEta := val[0].(int), val[1].([]byte), val[2].([]byte)
		if !bytes.Equal(storedEta, eta) {
			return result
		}

		for addr != nil {
			fernetKey := fernet.Key(key)
			encryptedNode := sse.ArrayA[addr]
			decryptedNode, err := fernetKey.Decrypt(encryptedNode)
			if err != nil {
				log.Fatal(err)
			}

			node := map[string]interface{}
			_ = json.Unmarshal(decryptedNode, &node)
			result = append(result, node["doc_id"].(string))

			addr = node["next_addr"]
			key = node["next_key"]
		}
	}
	return result
}

// 从文件中构建文档索引
func BuildDocumentsFromFiles() map[string]map[string]bool {
	documents := make(map[string]map[string]bool)
	documentsFolder := "D:\\study\\python-learn\\Big_Data\\file_test2"

	files, err := ioutil.ReadDir(documentsFolder)
	if err != nil {
		log.Fatal(err)
	}

	for _, file := range files {
		if !strings.HasSuffix(file.Name(), ".txt") {
			continue
		}

		content, err := ioutil.ReadFile(documentsFolder + "\\" + file.Name())
		if err != nil {
			log.Println("读取文件错误:", err)
			continue
		}

		keywords := ExtractKeywords(string(content))
		for _, keyword := range keywords {
			if _, exists := documents[keyword]; !exists {
				documents[keyword] = make(map[string]bool)
			}
			documents[keyword][file.Name()] = true
		}
	}

	return documents
}

// 从文本内容中提取关键词
func ExtractKeywords(content string) []string {
	// 这里只做示例，可以扩展为你自己的TF-IDF或TextRank方法
	return []string{"example", "keyword", "test"} // 这里只是返回示例
}

func main() {
	// 从文件构建文档集合
	documents := BuildDocumentsFromFiles()

	if len(documents) == 0 {
		fmt.Println("没有成功构建文档索引")
		return
	}

	// 初始化并运行SSE
	sse := NewSSE()
	sse.Setup()

	// 构建索引
	sse.BuildIndex(documents)

	for {
		var keyword string
		fmt.Print("请输入要搜索的关键词 (输入 'q' 退出): ")
		fmt.Scanln(&keyword)
		if keyword == "q" {
			break
		}

		// 生成陷门并搜索
		trapdoor := sse.GenTrapdoor(keyword)
		result := sse.Search(trapdoor)

		if len(result) > 0 {
			fmt.Printf("\n找到以下文件包含关键词 '%s':\n", keyword)
			for i, file := range result {
				fmt.Printf("%d. %s\n", i+1, file)
			}
		} else {
			fmt.Printf("\n未找到包含关键词 '%s' 的文件\n", keyword)
		}
	}
}

File: 大数据隐私保护/dsjclass02/main.py
import os
import random
from cryptography.fernet import Fernet
from typing import Dict, List, Set, Tuple
import jieba
import jieba.analyse
import hashlib
import logging
import struct
from collections import defaultdict

jieba.setLogLevel(logging.INFO)

class SSE:
    def __init__(self):
        self.K1 = None  # 用于伪随机置换的密钥
        self.K2 = None  # 用于伪随机函数的密钥
        self.K3 = None  # 用于伪随机置换的密钥
        self.array_A = {}  # 存储加密节点的数组
        self.table_T = {}  # 查找表

    def setup(self):
        """生成密钥"""
        self.K1 = Fernet.generate_key()
        self.K2 = Fernet.generate_key()
        self.K3 = Fernet.generate_key()
        return self.K1, self.K2, self.K3

    def pseudo_random_perm(self, key: bytes, value: str) -> int:
        """伪随机置换函数"""
        combined = key + value.encode()
        return int.from_bytes(hashlib.sha256(combined).digest()[:4], 'big') % 10000

    def pseudo_random_func(self, key: bytes, value: str) -> bytes:
        """伪随机函数"""
        combined = key + value.encode()
        return hashlib.sha256(combined).digest()

    def build_index(self, documents: Dict[str, Set[str]]):
        """构建索引"""
        ctr = 1
        for keyword, doc_ids in documents.items():
            doc_ids = list(doc_ids)
            Ki_prev = Fernet.generate_key()

            gamma = self.pseudo_random_perm(self.K3, keyword)
            eta = self.pseudo_random_func(self.K2, keyword)

            first_addr = self.pseudo_random_perm(self.K1, str(ctr))
            self.table_T[gamma] = (first_addr, Ki_prev, eta)

            for j, doc_id in enumerate(doc_ids):
                Ki_curr = Fernet.generate_key() if j < len(doc_ids) - 1 else None
                next_addr = self.pseudo_random_perm(self.K1, str(ctr + 1)) if j < len(doc_ids) - 1 else None

                node = {
                    'doc_id': doc_id,
                    'next_key': Ki_curr,
                    'next_addr': next_addr
                }
                f = Fernet(Ki_prev)
                encrypted_node = f.encrypt(str(node).encode())
                self.array_A[self.pseudo_random_perm(self.K1, str(ctr))] = encrypted_node

                Ki_prev = Ki_curr
                ctr += 1

    def gen_trapdoor(self, keyword: str) -> Tuple[int, bytes]:
        """生成陷门"""
        gamma = self.pseudo_random_perm(self.K3, keyword)
        eta = self.pseudo_random_func(self.K2, keyword)
        return (gamma, eta)

    def search(self, trapdoor: Tuple[int, bytes]) -> List[str]:
        """搜索文档"""
        gamma, eta = trapdoor
        result = []

        if gamma not in self.table_T:
            return result

        addr, key, stored_eta = self.table_T[gamma]

        # 字节串比较
        if stored_eta != eta:
            return result

        while addr is not None:
            try:
                f = Fernet(key)
                encrypted_node = self.array_A[addr]
                decrypted_node = f.decrypt(encrypted_node).decode()
                node = eval(decrypted_node)
                result.append(node['doc_id'])
                addr = node['next_addr']
                key = node['next_key']
            except Exception as e:
                break

        return result

class DocumentManager:
    def __init__(self, documents_folder: str):
        self.documents_folder = documents_folder
        self.documents = defaultdict(set)

    def extract_keywords(self, content: str, topK=20) -> List[str]:
        """从文本中提取关键词"""
        keywords = set()
        try:
            # 使用 TF-IDF 和 TextRank 方法
            keywords.update(jieba.analyse.extract_tags(content, topK=topK))
            keywords.update(jieba.analyse.textrank(content, topK=topK))
        except Exception as e:
            print(f"关键词提取过程中发生错误：{str(e)}")
        return list(keywords)

    def build_documents(self) -> Dict[str, Set[str]]:
        """从文件中构建文档集合"""
        if not os.path.exists(self.documents_folder):
            print(f"文件夹 {self.documents_folder} 不存在")
            return {}

        for filename in os.listdir(self.documents_folder):
            if not filename.endswith('.txt'):
                continue

            try:
                with open(os.path.join(self.documents_folder, filename), 'r', encoding='utf-8') as f:
                    content = f.read().strip()
                    if not content:
                        print(f"文件 {filename} 为空")
                        continue

                    keywords = self.extract_keywords(content)
                    if not keywords:
                        print(f"从文件 {filename} 中没有提取到关键词")
                        continue

                    for keyword in keywords:
                        self.documents[keyword].add(filename)

            except Exception as e:
                print(f"处理文件 {filename} 时发生错误：{str(e)}")

        return self.documents

def main():
    documents_folder = r"test"  # 文件夹路径
    document_manager = DocumentManager(documents_folder)

    # 从文件构建文档集合
    documents = document_manager.build_documents()

    if not documents:
        print("没有成功构建文档索引")
        return

    # 初始化并运行SSE
    sse = SSE()
    sse.setup()

    # 构建索引
    sse.build_index(documents)

    while True:
        keyword = input("请输入要搜索的关键词 (输入 'q' 退出): ")
        if keyword.lower() == 'q':
            break

        # 生成陷门并搜索
        trapdoor = sse.gen_trapdoor(keyword)
        result = sse.search(trapdoor)

        if result:
            print(f"\n找到以下文件包含关键词 '{keyword}':")
            for i, file in enumerate(result, 1):
                print(f"{i}. {file}")
        else:
            print(f"\n未找到包含关键词 '{keyword}' 的文件")

if __name__ == "__main__":
    main()
